src/lib/memory/Mem.cc

```diff
@@ -1,45 +1,88 @@
-  size = sz;
-};
+DataPacket::DataPacket(uint64_t address, uint64_t size, DataPacketType type,
+                       uint64_t reqId, std::vector<char> data, bool fault)
+    : address_(address),
+      size_(size),
+      type_(type),
+      id_(reqId),
+      data_(data),
+      inFault_(fault) {}
 
-ReadRespPacket* ReadPacket::makeResponse(uint64_t bytesRead, char* data) {
-  return new ReadRespPacket(this->id, bytesRead, data);
-}
+DataPacket::DataPacket(const DataPacket& packet)
+    : address_(packet.address_),
+      size_(packet.size_),
+      type_(packet.type_),
+      id_(packet.id_),
+      data_(packet.data_),
+      inFault_(packet.inFault_) {}
+
+DataPacket::DataPacket(DataPacket&& packet)
+    : address_(std::exchange(packet.address_, 0)),
+      size_(std::exchange(packet.size_, 0)),
+      type_(std::exchange(packet.type_, NONE)),
+      id_(std::exchange(packet.id_, 0)),
+      data_(std::move(packet.data_)),
+      inFault_(std::exchange(packet.inFault_, 0)) {}
 
-ReadRespPacket::ReadRespPacket(uint64_t req_id, size_t bytes_read, char* dt)
-    : DataPacket(READ) {
-  reqId = req_id;
-  bytesRead = bytes_read;
-  data = dt;
+DataPacket& DataPacket::operator=(const DataPacket& packet) {
+  address_ = packet.address_;
+  size_ = packet.size_;
+  type_ = packet.type_;
+  id_ = packet.id_;
+  data_ = packet.data_;
+  inFault_ = packet.inFault_;
+  return *this;
 }
 
-WritePacket::WritePacket(uint64_t addr, size_t sz, const char* dt)
-    : DataPacket(WRITE), data(dt) {
-  address = addr;
-  size = sz;
-  data = dt;
+DataPacket& DataPacket::operator=(DataPacket&& packet) {
+  address_ = std::exchange(packet.address_, 0);
+  size_ = std::exchange(packet.size_, 0);
+  type_ = std::exchange(packet.type_, NONE);
+  id_ = std::exchange(packet.id_, 0);
+  data_ = std::move(packet.data_);
+  inFault_ = std::exchange(packet.inFault_, 0);
+  return *this;
 }
 
-WriteRespPacket* WritePacket::makeResponse(uint64_t bytesReturned) {
-  return new WriteRespPacket(this->id, bytesReturned);
+DataPacket DataPacket::makeIntoReadResponse(std::vector<char> data) {
+  // If type of DataPacket isn't READ_REQUEST return faulty DataPacket.
+  if (type_ != READ_REQUEST) {
+    std::cerr << "[SimEng::DataPacket] Cannot change DataPacket type to "
+                 "READ_RESPONSE as the request type isn't READ_REQUEST."
+              << std::endl;
+    return DataPacket(true);
+  }
+  type_ = READ_RESPONSE;
+  data_ = data;
+  inFault_ = false;
+  return *this;
 }
 
-WriteRespPacket::WriteRespPacket(uint64_t req_id, size_t bytes_written)
-    : DataPacket(WRITE) {
-  reqId = req_id;
-  bytesWritten = bytes_written;
+DataPacket DataPacket::makeIntoWriteResponse() {
+  // If type of DataPacket isn't WRITE_REQUEST return faulty DataPacket.
+  if (type_ != WRITE_REQUEST) {
+    std::cerr << "[SimEng::DataPacket] Cannot change DataPacket type to "
+                 "WRITE_RESPONSE as the request type isn't WRITE_REQUEST."
+              << std::endl;
```

Comment:
Fro both cerr prefixes, only a single `:` should be used

Reply:
done